### PR TITLE
Added new module win_disk_facts

### PR DESCRIPTION
**Pull request by marqelme**
##### SUMMARY
This change adds new Windows Ansible module win_disk_facts.

With the module you can retrieve and output detailed information about the attached disks on the target.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/windows/win_disk_facts.ps1

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```

##### ADDITIONAL INFORMATION
You can use this module in combination with the win_disk_management module in order to retrieve the disks status on the target.

```yaml
- name: get disk facts
  win_disk_facts:

- name: output first disk size
  debug:
    var: ansible_facts.disks[0].size

- name: get disk facts
  win_disk_facts:

- name: output second disk serial number
  debug:
    var: ansible_facts.disks[0].serial_number
```

No changes because the module is a new module.

### CODE REVIEW

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: I don't think we still add that specific line to files. I noticed it was being removed when updating the copyright/license statement earlier during a rewrite.

Comment (dagwieers) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Maybe you want to add your copyright statement (as you did with the documentation) ?

**Review by dagwieers — APPROVED** [review]
LGTM

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Thanks!
Ok, I have changed the header as you have proposed.

I would like to add some information about the partitions and volumes of the disks to the output.
Is this a good idea or not needed?

**Comment by marqelme** [comment]
@dagwieers @jborean93 
I have added now information about the partitions, volumes and physical disk.
If this is ok I would update the .py docu with it and implement some more error handling in PowerShell file.

The Shippable tests failed now because there was no physical disk MediaType property available for Windows 2012 test. But I don't wanted to support 2012, it should work from 2012R2 on.
How can I pass this to the shippable tests?

**Comment by marqelme** [comment]
@dagwieers 
What I have to to do to get this further verified here?

**Comment by dagwieers** [comment]
@marqelme I think it is looking fine. Maybe add your PR to the Windows WG meeting agenda so a decision can be made ? cc: @nitzmahone @jborean93 @jhawkesworth

**Comment by marqelme** [comment]
@dagwieers 
Thank you, added it to the agenda.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: I would check to make sure `Get-Disk is available on the OS, IIRC it was added in Server 2012/Windows 8.

See https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/windows/win_region.ps1#L250 on how this can be done in Powershell. If the cmdlet isn't available we should fail with an exception message that tells us why.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: Is there a reason why we cast this as a string, we should be returning it as an integer on the return value.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: Instead of having a dict as a return value we should return a list of dict entries. What I would do is

set `$result.ansible_facts.ansible_disk = @()` in the original definition. In each loop create a disk of results like `$disk_results = @{}` and at the end of the loop add that like `$result.ansible_facts.ansible_disk += $disk-results`

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: This should probably be worded in the opposite way, you are looping through `$disks` to get a `$disk` not the other way round.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: Same for partitions, each `$disk_result` should be a list of partitions and not a dict with arbitrary keys.

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.py]: Until the latter is merged I would avoid referencing it in the module docs.

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.py]: You can enclose a URL in `U(https://...)`, for this to work you would need to also remove the `(v=vs.85)` from the URL.

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.py]: You don't need to document `changed` and `msg` as they are generic return values.

**Review by jborean93 — CHANGES_REQUESTED** [review]
Looks great, just a few comments

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Implemented

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Changed

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Absolutely right, changed that

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Implemented it in this way now

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: same as for disks

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.py]: removed it

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.py]: changed it

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.py]: removed

**Comment by marqelme** [comment]
Hey @jborean93 , I have implemented all the requested changes.
Could you please check it again?

**Review by dagwieers — APPROVED** [review]
LGTM

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: You are return different types of values (disk count and disk info) so this should still be a hashtable. While I said we should return a list in my original comments, I meant that the information about each disk should be in a list. If there are any extra values based on disks together this should be outside that list of disk info. For example this should probably look like

```
"ansible_facts": {
    "total_disks": 1,
    "disks": [
        {
            "bootable": true,
            "bus_type": "ATA",
            "clustered": false,
            ...
        }, ...
    ]
}
```

To get this you would do

```
$ansible_disks = @{
    total_disks = 0
    disks = @()
}
$results = @{
    changed = $false
    ansible_facts = @{
        $ansible_disks
    }
}
...

$ansible_disks.total_disks = $diskcount

...
# loop through each disk
$disk_info = @{}
$disk_info.bus_type = ...
$disk_info.clustered = ...
$ansible_disks.disks += $disk_info
```

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: I'm wondering whether this should just the size in bytes without any unit prefix to make things simple. The benefit of this approach is that you have a raw integer in the variable key that people can now easily perform operations on. The tradeoff is GB and TB values will be quite large, what are your thoughts?

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: When this is null, the drive letter is coming up as `\u0000` (`0x00`), if you remove the casting to a string this return `null` in the JSON result which is better.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: Will this ever happen, is it when you run as a non-administrator?

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: `Get-Disk` is more the OS restriction, the WMI namespaces that drive that cmdlet were only added in Windows 8/Server 2012. You can get rid of the `Get-Disk` check and reduce the indentation by adding `#AnsibleRequires -OSVersion 6.2` after your `#Requires -Module` section and Ansible will do the checking for you.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: s/byte/bytes/ but this depends on whether we want to include the units on the values.

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.py]: I would have thought the only restriction would be Windows 8.1 / Windows Server 2012 (NT 6.2), it works when I try it locally.

On a side note technically this should also work on Windows 8 but we don't support that OS so it should be kept at 8.1.

File: test/integration/targets/win_disk_facts/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_disk_facts/tasks/main.yml]: You should also have a test for a failure when `Get_Disk` is not available to ensure it fails with an error that makes sense.

File: test/integration/targets/win_disk_facts/tasks/tests.yml
Comment (jborean93) [inline test/integration/targets/win_disk_facts/tasks/tests.yml]: Can you add a generic `- disks_found.ansible_disks.disks[0].bootable is defined` for some of the important return values. You don't have to do them all, just some.

**Review by jborean93 — COMMENTED** [review]
Looks good, hopefully this is the last set of comments. I think the output format still needs to be modified a bit, I've put an example in the comments. The only major thing I think we should change is to remove the units on the return values. I've put my thoughts in one of the comments but interested in what you think.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Yes, now I got what you mean. I have changed this, now it makes sense and looks fine.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: In my opinion we should use the formatted approach (rounded and with prefix) if the module is only used to display the disk facts. But if it the values will be used for further calculations or another module we should remove the prefix and rounding.
What do you think?

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: I have implemented that, but would be removed depending on above decision.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Ok, thought it is better to have an empty string than null.
I have changed this now.

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Removed it

File: lib/ansible/modules/windows/win_disk_facts.ps1
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.ps1]: Implemented this

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.py]: Could you please test this in Win2012 with the virtual disk properties (you would need a Stroage Space and create a virtual disk there)?
Because I am not sure whether all virtual disk properties were available in Win2012.

File: test/integration/targets/win_disk_facts/tasks/tests.yml
Comment (marqelme) [reply test/integration/targets/win_disk_facts/tasks/tests.yml]: I have added some of the properties

File: test/integration/targets/win_disk_facts/tasks/main.yml
Comment (marqelme) [reply test/integration/targets/win_disk_facts/tasks/main.yml]: I have implemented a check which is throwing a message now.
The integration test seems to work because it is throwing the message in shippable tests for Win2008R2:

2017-12-15 14:25:26 fatal: [windows-2008-SP2]: FAILED! => {
2017-12-15 14:25:26     "changed": false, 
2017-12-15 14:25:26     "msg": "Cmdlet Get-Disk is not available in your Windows version"
2017-12-15 14:25:26 }

But this also cancels the shippable test with an error.
How can I avoid this?
I thought this could be avoided with #AnsibleRequires -OSVersion 6.2 ?

Comment (nwsparks) [reply test/integration/targets/win_disk_facts/tasks/main.yml]: You can add "failed_when:" and define it to only fail when the "msg" doesn't match the expected result for a server missing the module. See an example here: https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_power_plan/tasks/main.yml

It also shows how to skip running the tests on server versions that don't support it.

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (marqelme) [reply lib/ansible/modules/windows/win_disk_facts.py]: One additional thing:
I have added a check now which OS version is used:
`if([single]"$([System.Environment]::OSVersion.Version.Major).$([System.Environment]::OSVersion.Version.Minor)" -ge 6.3) {`

But would it not be better to check against the PowerShell version?

File: test/integration/targets/win_disk_facts/tasks/main.yml
Comment (marqelme) [reply test/integration/targets/win_disk_facts/tasks/main.yml]: Yes, thank you! But it results in the same...

Comment (jborean93) [reply test/integration/targets/win_disk_facts/tasks/main.yml]: I’ve done it in a few tests, have a look at the win_dsc ones as well.

Comment (marqelme) [reply test/integration/targets/win_disk_facts/tasks/main.yml]: I have implemented it.

File: lib/ansible/modules/windows/win_disk_facts.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_disk_facts.py]: 2012 comes with Powershell 3 which seems to work fine with the code so far so the restriction isn't there. I can't test with storage spaces but that should not be a reason why we would restrict the version.

**Review by jborean93 — APPROVED** [review]

**Comment by jborean93** [comment]
I'm pretty happy with the changes so far, my only hesitation is around how the sizes are displayed with the units. I'll check internally and see what some of the others think.

**Comment by jborean93** [comment]
@marqelme after discussing it internally it was said to keep the size return values with the unit suffix so it is more human readable. One thing that was pointed out was to make sure the prefix was correct and it showed whether the value was a gigabyte (GB) or gibibyte (GiB). In the case of PowerShell `1 * 1GB == 1073741824` which is in fact 1GiB and not 1GB (1000000000 bytes). Are you able to update the suffix for these return values to be `GiB`, `MiB`, `KiB` wherever it is needed?

In the future we could always add another option to the module that can modify these outputs according to the value set like so

```
- win_disk_facts:
    size_format: GiB
```

When not omitted it will use the defaults but let's not hold this merge up for this it's just food for thought.

**Comment by marqelme** [comment]
@jborean93 , this is a little confusing in PS.
 But you are absolutely right and I have changed that in the script and documentation.

The option to set the size format and maybe to have an option to show not all information I would prefer to implement in the next release.

**Comment by jborean93** [comment]
Awesome, perfectly fine to introduce in the next release if you want or if there is demand for it :). The whole gibibytes and gigabytes thing is just misleading but important to get right. Thanks for making the changes. Are you able to add this to the changelog as a new module when it is merged?

rebuild_merge

**Comment by marqelme** [comment]
Yes sure, you mean add it to changelog.md in my fork and merge it then?

**Comment by jborean93** [comment]
Just created a new PR with the change log update.

**Comment by marqelme** [comment]
@jborean93 Thanks for the merge! :)
Please see above pull request #34087 regarding CHANGELOG

**Pull request merged**
